pkg/prometheus/statefulset.go

```diff
@@ -594,38 +594,43 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
 		})
 	}
 
-	const localProbe = `if [ -x "$(command -v curl)" ]; then exec curl %s; elif [ -x "$(command -v wget)" ]; then exec wget -q -O /dev/null %s; else exit 1; fi`
-
-	var readinessProbeHandler v1.Handler
-	{
-		readyPath := path.Clean(webRoutePrefix + "/-/ready")
+	probeHandler := func(probePath string) v1.Handler {
+		probePath = path.Clean(webRoutePrefix + probePath)
+		handler := v1.Handler{}
 		if p.Spec.ListenLocal {
-			localReadyPath := fmt.Sprintf("http://localhost:9090%s", readyPath)
-			readinessProbeHandler.Exec = &v1.ExecAction{
+			handler.Exec = &v1.ExecAction{
 				Command: []string{
 					"sh",
 					"-c",
-					fmt.Sprintf(localProbe, localReadyPath, localReadyPath),
+					fmt.Sprintf(`if [ -x "$(command -v curl)" ]; then exec curl %[1]s; elif [ -x "$(command -v wget)" ]; then exec wget -q -O /dev/null %[1]s; else exit 1; fi`, fmt.Sprintf("http://localhost:9090%s", probePath)),
 				},
 			}
 		} else {
-			readinessProbeHandler.HTTPGet = &v1.HTTPGetAction{
-				Path: readyPath,
+			handler.HTTPGet = &v1.HTTPGetAction{
+				Path: probePath,
 				Port: intstr.FromString(p.Spec.PortName),
 			}
 			if p.Spec.Web != nil && p.Spec.Web.TLSConfig != nil && version.GTE(semver.MustParse("2.24.0")) {
-				readinessProbeHandler.HTTPGet.Scheme = v1.URISchemeHTTPS
+				handler.HTTPGet.Scheme = v1.URISchemeHTTPS
 			}
 		}
+		return handler
+	}
+
+	startupProbe := &v1.Probe{
+		Handler:          probeHandler("/-/healthy"),
```

Comment:
Late to the party but my understanding is that `/-/healthy` will return OK as soon as the web server is started (independently of the WAL replay). I assume that we should use `/-/ready` which returns OK only after the TSDB has been initialized.

Reply:
@simonpasquier I think this PR and its use of startup probe was correct.

> The kubelet uses startup probes to know when a container application has started. If such a probe is configured, it disables liveness and readiness checks until it succeeds, making sure those probes don't interfere with the application startup. 

I would think a readiness probe is more fitting for the `ready` endpoint?

Reply:
When Prometheus restarts, it will concurrently start the web server and replay the WAL directory (which can take up to several minutes).

* `/-/healthy` returns 200 OK as soon as the web server is ready to answer requests independently of the WAL replay so it should happen quite fast.
* `/-/ready` returns 503 Service Unavailable while Prometheus loads the TSDB data and 200 OK once it has completed.

IIUC the [purpose](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) of the startup probe is to ensure that the container has completed its initialization so waiting for the TSDB to be initialized seems the right thing to do.

IIRC a correlated issue was that WAL replay is really CPU intensive and the liveness probe could time out during startup triggering a restart of the container (which is just making the situation even worse).

Reply:
I don’t have a strong opinion on this:
- on one hand, I tend to agree with you that `/-/ready` seems more appropriate to signify startup is done,
- on the other hand, if using `/-/ready`, if for any reason the startup process hangs, the pod won’t have a chance to be restarted by Kubernetes (though one could argue that this will be the case anyway until we restore the liveness probe).

Reply:
to be clear I think that using the `/-/healthy` handler for the startup probe will cause issues for Prometheus instances that have long WAL replays: the startup probe would return ok almost immediately, the readiness probe kicks in but it will fail if the TSDB doesn't initialize in less than 15 seconds (= period of 5s x 3 failures).

Reply:
IIUC this implementation should be fine on it's own no? If the readiness probe fails while the TSDB is not initialised then traffic won't be routed to the Pod but the Pod won't be restarted. We might begin to get issues when we add a liveness probe. See my comment https://github.com/prometheus-operator/prometheus-operator/pull/4387#issuecomment-984455671 which is why I agree we should now change this implementation.

Reply:
I agree with both of you (see my comment in https://github.com/prometheus-operator/prometheus-operator/pull/4387#issuecomment-984584292).